minorminer/utils/feasibility.py

```diff
@@ -0,0 +1,222 @@
+                # marginal degree distribution filter is used.
+
+                # We can eliminate nodes of equal degree assuming best case:
+                ResidualCounts = Counter(T_degree)
+                ResidualCounts.subtract(Counter(S_degree))
+
+                # Target nodes of degree x <=2 are only of use for minor
+                # embedding source nodes of degree <=x:
+                for kS in range(3):
+                    if ResidualCounts[kS] < 0:
+                        ResidualCounts[kS + 1] += ResidualCounts[kS]
+                    ResidualCounts[kS] = 0
+
+                if all(v > 0 for v in ResidualCounts.values()):
+                    return True
+                nT_auxiliary = sum(ResidualCounts.values())
+                if nT_auxiliary < 0:  # extra available to form chains
+                    return False
+
+                # In best case all target nodes have degree kTmax, and chains
+                # are trees. To cover degree k in S requires n auxiliary target
+                # nodes such that kTmax + n(kTmax-2) >= k
+                kTmax = np.max([k for k, v in ResidualCounts.items() if v > 0])
+                min_auxiliary_necessary = sum(
+                    [
+                        -v * np.ceil((k - kTmax) / (kTmax - 2))
+                        for k, v in ResidualCounts.items()
+                        if v < 0
+                    ]
+                )
+                return min_auxiliary_necessary <= nT_auxiliary
+        else:
+            return True
+
+
+def lattice_size_upper_bound(T: nx.Graph = None) -> int:
+    """Determines an upper bound on the lattice size.
+
+    The lattice size is the parameter ``m`` of a dwave_networkx graph, also
+    called number of rows.
+
+    Args:
+        T: The target graph in which to embed. The graph must be of type
+            zephyr, pegasus or chimera and constructed by dwave_networkx.
+    Returns:
+        int: The maximum possible size of a tile
+    """
+    return max(T.graph.get("rows"), T.graph.get("columns"))
+
+
+def lattice_size_lower_bound(
+    S: nx.Graph,
+    T: nx.Graph = None,
+    topology: str = None,
+    t: int = None,
+    one_to_one: bool = False,
+) -> float:
+    """Returns a lower bound on the size necessary for embedding.
+
+    The lattice size is the parameter `m` of a dwave_networkx graph, also
+    called number of rows. The function returns a lower bound (necessary but
+    not sufficient for embedding) using efficiently established graph
+    properties such as the number of nodes, number of edges, node-degree
+    distribution, and two-colorability
+
+    Args:
+        S: The source graph to embed.
+        T: The target graph in which to embed. The graph must be of type
+            'zephyr', 'pegasus', or 'chimera' and constructed by dwave_networkx.
+        topology: The topology ('chimera', 'pegasus', or 'zephyr'). This is
+            inferred from `T` by default. Any set value must be consistent with
```

Comment:
```suggestion
            inferred from ``T`` by default. Any set value must be consistent with
```
Same for other cases in docstring.